In [ ]:
# Self Notes:
# 1) Sec III - Answer number 5!
# 2) Flattten matrix results function (and seperate it and generalize it in the future!)

In [1]:
import pandas as pd
import numpy as np
import seaborn as sns

import warnings
warnings.filterwarnings("ignore")

from sklearn.tree import DecisionTreeClassifier, plot_tree
from sklearn.metrics import classification_report
from sklearn.metrics import confusion_matrix

import matplotlib.pyplot as plt
import seaborn as sns

from acquire import get_titanic_data
from prepare import prep_titanic
from splitter import splitter

# I. Decision Tree - Titanic Data

In [2]:
df = get_titanic_data()
df.head(2)

passenger_id  survived  pclass     sex   age  sibsp  parch     fare  \
0             0         0       3    male  22.0      1      0   7.2500   
1             1         1       1  female  38.0      1      0  71.2833   

  embarked  class deck  embark_town  alone  
0        S  Third  NaN  Southampton      0  
1        C  First    C    Cherbourg      0

In [ ]:
df = prep_titanic(df)
df.drop(columns='age', inplace=True)
df.head(2)

## 1. What is your baseline prediction?  Baseline accuracy?

In [ ]:
df_train, df_validate, df_test = splitter(df, 'survived')
df_train.survived.value_counts()

In [ ]:
307/(307+191)

#### Baseline prediction is died, which is accurate 61.65% of the time.

## 2. Fit the decision tree classifier to your training sample and transform (i.e. make predictions on the training sample)

In [ ]:
X_train = df_train.drop(columns=['survived'])
y_train = df_train.survived

X_validate = df_validate.drop(columns=['survived'])
y_validate = df_validate.survived

X_test = df_test.drop(columns=['survived'])
y_test = df_test.survived

In [ ]:
clf = DecisionTreeClassifier(max_depth=3, random_state=123)
clf = clf.fit(X_train, y_train)
# this is tpyiall where y_pred = clf.predict(X_train) would live as part of the "3 lines of code"

In [ ]:
# plotting figure unecessary but good for funsies or stakeholder visuals at times
plt.figure(figsize=(16, 8))
plot_tree(clf, feature_names=X_train.columns, class_names=clf.classes_.astype(str), rounded=True)

## 3. Evaluate your in-sample results using the model score, confusion matrix, and classification report.

In [ ]:
y_pred = clf.predict(X_train)
y_pred[10:15]

In [ ]:
y_pred_proba = clf.predict_proba(X_train)
y_pred_proba[10:15]

In [ ]:
# unecessary/redundant - found in dataframe you create below already
confusion_matrix(y_train, y_pred)

In [ ]:
pd.DataFrame(confusion_matrix(y_train, y_pred), index=['died','survived'], columns=['died','survived'])

## 4. Compute: Accuracy, true positive rate, false positive rate, true negative rate, false negative rate, precision, recall, f1-score, and support.

##### In sample:

In [ ]:
print('Accuracy of Decision Tree classifier on training set: {:.2f}'
      .format(clf.score(X_train, y_train)))

In [ ]:
report = classification_report(y_train, y_pred)
print(report)

##### Validate:

In [ ]:
print('Accuracy of Decision Tree classifier on validate set: {:.2f}'
     .format(clf.score(X_validate, y_validate)))

In [ ]:
y_pred = clf.predict(X_validate)

print(classification_report(y_validate, y_pred))

In [ ]:
# Make function that creates a fun and pretty table when you have time.

## 5. Run through steps 2-4 using a different max_depth value.

### Using max_depth = 7:

##### Train:

In [ ]:
clf = DecisionTreeClassifier(max_depth=7, random_state=123)
clf = clf.fit(X_train, y_train)

In [ ]:
plt.figure(figsize=(16, 8))
plot_tree(clf, feature_names=X_train.columns, class_names=clf.classes_.astype(str), rounded=True)

In [ ]:
print('Accuracy of Decision Tree classifier on training set: {:.2f}'
      .format(clf.score(X_train, y_train)))

In [ ]:
y_pred = clf.predict(X_train)
print(classification_report(y_train, y_pred))

##### Validate:

In [ ]:
print('Accuracy of Decision Tree classifier on validation set: {:.2f}'
      .format(clf.score(X_validate, y_validate)))

In [ ]:
y_pred = clf.predict(X_validate)

print(classification_report(y_validate, y_pred))

## 6. Which model performs better on your in-sample data?

The max_steps = 7 is better on in-sample data (as expected, more granular).

## 7. Which model performs best on your out-of-sample data, the validate set?

The max_steps = 3 is better on the validate data.

# II. Decision Tree - Telco Data

In [ ]:
from acquire import get_telco_data
from prepare import prep_telco

In [ ]:
df = get_telco_data()
df.head(2)

In [ ]:
df = prep_telco(df)
df.info()

In [ ]:
dropcols = [col for col in df.columns if df[col].dtype == 'object']
dropcols.remove('churn')
dropcols.append('churn_encoded')
dropcols

In [ ]:
df = df.drop(columns=dropcols)
df.head(2)

In [ ]:
df_train, df_validate, df_test = splitter(df, 'churn')
df_train.churn.value_counts()

In [ ]:
#Note: Work more on establishing the baseline with code.  Esp since often a categorical!
2891/(2891+1046)

#### Baseline case is 'not churned' and the baseline accuracy is 73.43%.

In [ ]:
#Note - you can also drop all the columns in X-data, including churn, in the drop(columns)
# The way you did it here was redundant!
X_train = df_train.drop(columns=['churn'])
y_train = df_train.churn

X_validate = df_validate.drop(columns=['churn'])
y_validate = df_validate.churn

X_test = df_test.drop(columns=['churn'])
y_test = df_test.churn

In [ ]:
clf = DecisionTreeClassifier(max_depth = 3, random_state=123)
clf = clf.fit(X_train, y_train)

In [ ]:
plt.figure(figsize=(16, 8))
plot_tree(clf, feature_names=X_train.columns, class_names=clf.classes_.astype(str), rounded=True)

In [ ]:
y_pred = clf.predict(X_train)
y_pred[0:5]

In [ ]:
y_pred_proba = clf.predict_proba(X_train)
y_pred_proba[0:5]

In [ ]:
labels = sorted(y_train.unique())
pd.DataFrame(confusion_matrix(y_train, y_pred), index=labels, columns=labels)

##### In training sample:

In [ ]:
print('Accuracy of Decision Tree classifier on training set: {:.2f}'
      .format(clf.score(X_train, y_train)))

In [ ]:
y_pred = clf.predict(X_train)
print(classification_report(y_train, y_pred))

##### In validation sample:

In [ ]:
print('Accuracy of Decision Tree classifier on validation set: {:.2f}'
      .format(clf.score(X_validate, y_validate)))

In [ ]:
y_pred = clf.predict(X_validate)
print(classification_report(y_validate, y_pred))

### Using max_depth = 7

In [ ]:
clf = DecisionTreeClassifier(max_depth = 7, random_state=123)
clf = clf.fit(X_train, y_train)

In [ ]:
plt.figure(figsize=(16, 8))
plot_tree(clf, feature_names=X_train.columns, class_names=clf.classes_.astype(str), rounded=True)

In [ ]:
y_pred = clf.predict(X_train)
y_pred[0:5]

In [ ]:
y_pred_proba = clf.predict_proba(X_train)
y_pred_proba[0:5]

In [ ]:
labels = sorted(y_train.unique())
pd.DataFrame(confusion_matrix(y_train, y_pred), index=labels, columns=labels)

##### In training sample:

In [ ]:
print('Accuracy of Decision Tree classifier on training set: {:.2f}'
      .format(clf.score(X_train, y_train)))

In [ ]:
y_pred = clf.predict(X_train)
print(classification_report(y_train, y_pred))

##### In validation sample:

In [ ]:
print('Accuracy of Decision Tree classifier on validation set: {:.2f}'
      .format(clf.score(X_validate, y_validate)))

In [ ]:
y_pred = clf.predict(X_validate)
print(classification_report(y_validate, y_pred))

### Comparison at different max levels:

In [ ]:
for i in range(1,9):
    clf = DecisionTreeClassifier(max_depth = i, random_state=123)
    clf = clf.fit(X_train, y_train)
    print(f'Depth = {i}')
    print('Accuracy of Decision Tree classifier on training set: {:.2f}'
      .format(clf.score(X_train, y_train)))
    print('Accuracy of Decision Tree classifier on validation set: {:.2f}'
      .format(clf.score(X_validate, y_validate)))

#### Model performs best on validation set at a Depth of 3.

In [ ]:
# Note: Remember you are not necessarily optimizing for accuracy, could be looking at recall or precision (or even F1)
# A good way to assess train v. validation is by minimizing the distance between the outcome of interest

# III. Random Forest - Titanic Data

##### Note: Used acquire, prep and splitting functions found in first exercise above.

## 1. Fit the Random Forest classifier to your training sample and transform (i.e. make predictions on the training sample) setting the random_state accordingly and setting min_samples_leaf = 1 and max_depth = 10.

In [ ]:
from sklearn.ensemble import RandomForestClassifier
rf = RandomForestClassifier(max_depth=10, 
                            min_samples_leaf = 1,
                            random_state=123)

In [ ]:
rf = rf.fit(X_train, y_train)

In [ ]:
print(X_train.columns)
print(rf.feature_importances_)

## 2. Evaluate your results using the model score, confusion matrix, and classification report.

### Training data:

In [ ]:
print('Accuracy of random forest classifier on training set: {:.2f}'
     .format(rf.score(X_train, y_train)))

In [ ]:
labels = ['died','survived']
y_pred_train = rf.predict(X_train)
dfrftr = pd.DataFrame(confusion_matrix(y_train, y_pred), index=labels, columns=labels)
dfrftr

In [ ]:
print(classification_report(y_train, y_pred))

In [ ]:
pd.DataFrame(classification_report(y_train, y_pred, output_dict=True))

### Validation Data:

In [ ]:
print('Accuracy of random forest classifier on training set: {:.2f}'
     .format(rf.score(X_validate, y_validate)))

In [ ]:
y_pred_validate = rf.predict(X_validate)
dfrfv = pd.DataFrame(confusion_matrix(y_validate, y_pred_validate), index=labels, columns=labels)
dfrfv

In [ ]:
print(classification_report(y_validate, y_pred_validate))

## 3. Print and clearly label the following: Accuracy, true positive rate, false positive rate, true negative rate, false negative rate, precision, recall, f1-score, and support.

In [22]:
def matrix_results(df, TP='NW', actual='rows'):
    if TP=='NW':
        if actual=='rows':
            true_positive = df.iloc[0,0]
            false_positive = df.iloc[1,0]
            true_negative = df.iloc[1,1]
            false_negative = df.iloc[0,1]
        elif actual=='columns':
            true_positive = df.iloc[0,0]
            false_positive = df.iloc[0,1]
            true_negative = df.iloc[1,1]
            false_negative = df.iloc[1,0]
        else:
            print('''If actual results (from training data) is in columns, please include "actual = 'columns'" in the function parameters''')
            return
    elif TP=='SE':
        if actual=='rows':
            true_positive = df.iloc[1,1]
            false_positive = df.iloc[0,1]
            true_negative = df.iloc[0,0]
            false_negative = df.iloc[1,0]
        elif actual=='columns':
            true_positive = df.iloc[1,1]
            false_positive = df.iloc[1,0]
            true_negative = df.iloc[0,0]
            false_negative = df.iloc[0,1]
        else:
            print('''If actual results (from training data) is in columns, please include " actual = 'columns' " in the function parameters''')
            return
    else:
        print('''If True Positive is in the bottom right quadrant, please include " TP = 'SE' " in the function parameters''')
        return
    precision = true_positive / (true_positive + false_positive)
    recall = true_positive / (true_positive + false_negative)
    accuracy = (true_positive + true_negative) / (true_positive + true_negative + false_positive + false_negative)
    f1_score = (2*precision*recall)/(precision + recall)
    support_pos = true_positive + false_negative
    support_neg = true_negative + false_positive
#     print(f'True Positive = {true_positive}')
#     print(f'False Positive = {false_positive}')
#     print(f'True Negative = {true_negative}')
#     print(f'False Negative = {false_negative}')
#     print(f'Precision = {precision:.2f}')
#     print(f'Recall = {recall:.2f}')
#     print(f'Accuracy = {accuracy:.2f}')
#     print(f'F1 Score = {f1_score:.2f}')
#     print(f'Support, Positive = {support_pos}')
#     print(f'Support, Negative = {support_neg}')
    return {'true_positive':true_positive, 'false_positive':false_positive, 'true_negative':true_negative, 'false_negative':false_negative, 'precision':precision, 'recall':recall, 'accuracy':accuracy, 'f1_score':f1_score, 'support_pos':support_pos, 'support_neg':support_neg}

In [ ]:
print('Training Data Results:')
matrix_results(dfrftr)
print('\n----------\n')
print('Validation Data Results:')
matrix_results(dfrfv)

In [ ]:
pd.DataFrame(matrix_results(dfrftr))

## 4. Run through steps increasing your min_samples_leaf and decreasing your max_depth.

In [ ]:
for i in range(10,0,-1):
    for j in range(1,10):
        rf = RandomForestClassifier(max_depth=i, min_samples_leaf = j,random_state=123)
        rf = rf.fit(X_train, y_train)
        y_pred_train = rf.predict(X_train)
        y_pred_validate = rf.predict(X_validate)
        print(f'Depth = {i}, Leaves = {j}')
        print('Accuracy of random forest classifier on training set: {:.2f}'.format(rf.score(X_train, y_train)))
        print('Accuracy of random forest classifier on validation set: {:.2f}'.format(rf.score(X_validate, y_validate)))
#         print('Training Report:')
#         print(classification_report(y_train, y_pred_train))
#         print('Validation Report:')
#         print(classification_report(y_validate, y_pred_validate))
#         print('\n----------\n')

## 5. What are the differences in the evaluation metrics? Which performs better on your in-sample data? Why?

In [ ]:
# ANSWER

# IV. KNN - Titanic Data

#### Acquired, prepped and split above in section I.
-Baseline is 'died' with an accuracy of 61.65%

## 1. Fit a K-Nearest Neighbors classifier to your training sample and transform (i.e. make predictions on the training sample)

In [ ]:
from sklearn.neighbors import KNeighborsClassifier
knn = KNeighborsClassifier()

In [ ]:
knn = knn.fit(X_train, y_train)

In [ ]:
y_pred = knn.predict(X_train)
y_pred_proba = knn.predict_proba(X_train)

## 2. Evaluate your results using the model score, confusion matrix, and classification report.

Model Score:

In [ ]:
knn.score(X_train, y_train)

Confusion Matrix:

In [ ]:
labels = ['died','survived']
results = pd.DataFrame(confusion_matrix(y_train, y_pred), index=labels, columns=labels)
results

In [ ]:
print(classification_report(y_train, y_pred))
pd.DataFrame(classification_report(y_train, y_pred, output_dict=True))

## 3. Print and clearly label the following: Accuracy, true positive rate, false positive rate, true negative rate, false negative rate, precision, recall, f1-score, and support.

In [ ]:
#pd.DataFrame(matrix_results(results).values(), columns=matrix_results(results).keys())
matrix_results(results)

## 4. Run through steps 2-3 setting k to 10

In [ ]:
knn = KNeighborsClassifier(n_neighbors=10)
knn = knn.fit(X_train, y_train)
y_pred = knn.predict(X_train)
knn.score(X_train, y_train)

In [ ]:
labels = ['died','survived']
results = pd.DataFrame(confusion_matrix(y_train, y_pred), index=labels, columns=labels)
results

In [ ]:
print(classification_report(y_train, y_pred))
pd.DataFrame(classification_report(y_train, y_pred, output_dict=True))

In [ ]:
matrix_results(results)

## 5. Run through steps 2-3 setting k to 20

In [ ]:
knn = KNeighborsClassifier(n_neighbors=20)
knn = knn.fit(X_train, y_train)
y_pred = knn.predict(X_train)
knn.score(X_train, y_train)

In [ ]:
labels = ['died','survived']
results = pd.DataFrame(confusion_matrix(y_train, y_pred), index=labels, columns=labels)
results

In [ ]:
print(classification_report(y_train, y_pred))
pd.DataFrame(classification_report(y_train, y_pred, output_dict=True)).round(2)

In [ ]:
matrix_results(results)

##### Trying to loop through all k for train data:

In [ ]:
metrics = []
labels = ['died','survived']

for i in range(1,25):
    knn = KNeighborsClassifier(n_neighbors=i, weights='uniform')
    knn = knn.fit(X_train, y_train)
    y_pred = knn.predict(X_train)
    
    result = pd.DataFrame(confusion_matrix(y_train, y_pred), index=labels, columns=labels)
    results = matrix_results(result)
    results['k'] = i
    
#     k = {'k':i}
#     k.update(results)
#     metrics.append(k)
    metrics.append(results)
df = pd.DataFrame(metrics).round(2).set_index('k')
df

## 6. What are the differences in the evaluation metrics? Which performs better on your in-sample data? Why?

#### It seems the lower the k, the better the model performs for overall accuracy.  Interestingly, if we are optimizing for recall (minimizing false negatives) k=2 is far superior.

## 7. Which model performs best on our out-of-sample data from validate?

##### Focusing on accuracy:

In [ ]:
differences=[]

for i in range(1,25):
    knn = KNeighborsClassifier(n_neighbors=i, weights='distance')
    knn = knn.fit(X_train, y_train)
    
    in_sample_acc = knn.score(X_train, y_train)
    out_sample_acc = knn.score(X_validate, y_validate)
    
    output = {
        'k' : i,
        'train_accuracy' : in_sample_acc,
        'validation_accuracy' : out_sample_acc
    }
    
    differences.append(output)

ddf = pd.DataFrame(differences).set_index('k')
ddf['difference'] = ddf['train_accuracy']-ddf['validation_accuracy']
ddf.sort_values(by='difference')

#### While the best k for train was very low, when appying it to the validation data, we see a higher k is better.  In this case k=11.

### Interesting - weights = distance overfits the training data!

# V. Logistic Regression - Titanic

## 1. Create a model that includes age in addition to fare and pclass. Does this model perform better than your baseline? 

In [3]:
from sklearn.linear_model import LogisticRegression
logit = LogisticRegression()

In [4]:
df = prep_titanic(df)
df.head()

survived  pclass   age  Siblings/Spouses  Parents/Children     fare  alone  \
0         0       3  22.0                 1                 0   7.2500      0   
1         1       1  38.0                 1                 0  71.2833      0   
2         1       3  26.0                 0                 0   7.9250      1   
3         1       1  35.0                 1                 0  53.1000      0   
4         0       3  35.0                 0                 0   8.0500      1   

   sex_male  embark_town_Queenstown  embark_town_Southampton  
0         1                       0                        1  
1         0                       0                        0  
2         0                       0                        1  
3         0                       0                        1  
4         1                       0                        1

In [7]:
df_1 = df[['survived', 'pclass','age','fare']]
df_1.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 891 entries, 0 to 890
Data columns (total 4 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   survived  891 non-null    int64  
 1   pclass    891 non-null    int64  
 2   age       714 non-null    float64
 3   fare      891 non-null    float64
dtypes: float64(2), int64(2)
memory usage: 34.8 KB


In [10]:
df_1 = df_1.dropna()

In [13]:
df_train, df_validate, df_test = splitter(df_1, 'survived')
df_train.survived.value_counts()

0    237
1    162
Name: survived, dtype: int64

In [25]:
237/(162+237)

0.5939849624060151

Baseline prediction for this subset of the data is "Dead" accurate 59.4% of the time.

In [14]:
X_train = df_train.drop(columns=['survived'])
y_train = df_train.survived

X_validate = df_validate.drop(columns=['survived'])
y_validate = df_validate.survived

X_test = df_test.drop(columns=['survived'])
y_test = df_test.survived

Fit the Logistical Regression Model to the Data

In [15]:
logit = logit.fit(X_train,y_train)

Feature and Intercept Examination 

In [20]:
print('Coefficient: \n', logit.coef_)
print('Intercept: \n', logit.intercept_)

Coefficient: 
 [[-1.19226315 -0.04138191  0.00146339]]
Intercept: 
 [3.42691873]


In [21]:
y_pred = logit.predict(X_train)
y_pred.mean()

0.3208020050125313

In [24]:
matrix_results(pd.DataFrame(confusion_matrix(y_train, y_pred)))

{'true_positive': 194,
 'false_positive': 77,
 'true_negative': 85,
 'false_negative': 43,
 'precision': 0.7158671586715867,
 'recall': 0.8185654008438819,
 'accuracy': 0.6992481203007519,
 'f1_score': 0.7637795275590552,
 'support_pos': 237,
 'support_neg': 162}

In [26]:
print(classification_report(y_train, y_pred))

              precision    recall  f1-score   support

           0       0.72      0.82      0.76       237
           1       0.66      0.52      0.59       162

    accuracy                           0.70       399
   macro avg       0.69      0.67      0.67       399
weighted avg       0.69      0.70      0.69       399



#### It is about 10% better (or 10/60% relatively better)